File: stack/minStack.py
class MinStack:

    def __init__(self):
        self.min = []
        self.stack = []

    def push(self, val: int) -> None:
        if len(self.stack) == 0:
            self.min.append(val)
            self.stack.append(val)
        else:
            if val < self.min[-1]:
                self.min.append(val)
                self.stack.append(val)
            else:
                self.min.append(self.min[-1])
                self.stack.append(val)

    def pop(self) -> None:
        self.stack.pop()
        self.min.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

if __name__ == "__main__":
    obj = MinStack()
    obj.push(-2)
    obj.push(0)
    obj.push(-3)
    print(obj.getMin())
    obj.pop()
    print(obj.top())
    print(obj.getMin())

File: stack/cppRealization/minStack.cpp
#include <iostream>
#include <stack>

using namespace std;

class Solution {
public:
    void push(int val) {
        if(stack_.empty()){
            min_.push(val);
            stack_.push(val);
        }else{
            if(val < min_.top()){
                min_.push(val);
                stack_.push(val);
            }else{
                min_.push(min_.top());
                stack_.push(val);
            }
        }
    }

    void pop() {
        stack_.pop();
        min_.pop();
    }

    int top() {
        return stack_.top();

    }

    int gitMin() {
        return min_.top();
    }

private:
    stack<int> stack_;
    stack<int> min_;
};


int main(){

    Solution obj;
    obj.push(-2);
    obj.push(0);
    obj.push(-3);
    cout << obj.gitMin() << endl;
    obj.pop();
    cout << obj.top() << endl;
    cout << obj.gitMin() << endl;
    return 0;
}